File: Arrays and String/Java/713. Subarray Product Less Than K.java
class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        if (k <= 1) {
            return 0;
        }

        int left = 0; 
        int ans = 0;
        int curr = 1;

        for (int right = 0; right < nums.length; right++) {
            
            curr = curr*nums[right];
            
            while (curr >= k) {
                curr = curr / nums[left];
                left += 1;
            }

            ans += (right - left + 1);
        }
        
        return ans;
    }
}

File: Arrays and String/Python/713. Subarray Product Less Than K.py
class Solution(object):
    def numSubarrayProductLessThanK(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        if k <= 1:
            return 0
    
        left = answer = 0
        curr = 1 # keep track of the product

        # open the windows to the right
        for right in range(len(nums)):
            curr *= nums[right]

            # if the product pass the limit (curr >= k)
            while curr >= k: 
                # shrink the window, update the product
                curr //= nums[left] 
                left += 1
            
            # count the numbers of the subarray ending with right
            # the number of valid windows ending at index right is equal to the size of the window, 
            # which we know is right - left + 1
            answer += (right - left + 1)

        # return the answer
        return answer
